SourceCode/CRTRenderer.cpp

```diff
@@ -2,103 +2,177 @@
         }
     }
-    return finalColor;
+    return CRTVector(clamp(finalColor.x, 0, 1), 
+        clamp(finalColor.y, 0, 1), 
+        clamp(finalColor.z, 0, 1));
+}
+
+CRTVector CRTRenderer::shadeReflective(const CRTRay& ray, const Intersection& data) const {
+    CRTVector normalVector = data.faceNormal;
+    const CRTMaterial& material = scene->getMaterial(data.materialIndex);
+    if (material.smoothShading) {
+        normalVector = data.smoothNormal;
+    }
+    CRTVector reflectedDirection = reflect(ray.direction, normalVector);
+    CRTRay reflectedRay{ data.hitPoint + normalVector * REFLECTION_BIAS, reflectedDirection, RayType::REFLECTIVE, ray.depth + 1 };
+    CRTVector uv = scene->getGeometryObject(data.hitObjectIndex).getUV(data);
+
+    return shade(reflectedRay, rayTrace(reflectedRay)) * scene->getTexture(material.textureName)->sample(uv.x, uv.y);
+}
+
+CRTVector CRTRenderer::shadeRefractive(const CRTRay& ray, const Intersection& data) const {
+    const CRTMaterial& material = scene->getMaterial(data.materialIndex);
+    float n1 = 1.0f; // Index of refraction of the air
+    float n2 = material.ior;
+    CRTVector normalVector = material.smoothShading ? data.smoothNormal : data.faceNormal;
+
+    float dotPr = dot(ray.direction, normalVector);
+    if (dotPr > 0) { // Ray is leaving the refractive object
+        normalVector *= -1;
+        dotPr *= -1;
+        std::swap(n1, n2);
+    }
+
+    float cosIncomming = -dotPr;
+    float sinIncomming = 1.0f - cosIncomming * cosIncomming;
+
+    if (sinIncomming >= ((n2 * n2) / (n1 * n1))) { // total internal reflection
+        CRTVector reflectedDirection = reflect(ray.direction, normalVector);
+        CRTRay reflectedRay{ data.hitPoint + normalVector * REFLECTION_BIAS, reflectedDirection, RayType::REFLECTIVE, ray.depth + 1 };
+        return shade(reflectedRay, rayTrace(reflectedRay));
+    }
+
+    float sinOutcomming = sqrt(sinIncomming) * n1 / n2;
+    float cosOutcomming = sqrt(1.0f - sinOutcomming * sinOutcomming);
+
+    CRTVector A = cosOutcomming * (-normalVector);
+    CRTVector C = ray.direction + (cosIncomming * normalVector);
+    C.normalize();
+    CRTVector refractedDirection = A + (C * sinOutcomming);
+    refractedDirection.normalize();
+
+    CRTRay refractedRay{ data.hitPoint + normalVector * -REFRACTION_BIAS, refractedDirection, RayType::REFRACTIVE, ray.depth + 1 };
+
+    CRTVector reflectedDirection = reflect(ray.direction, normalVector);
+    CRTRay reflectedRay{ data.hitPoint + normalVector * REFLECTION_BIAS, reflectedDirection, RayType::REFLECTIVE, ray.depth + 1 };
+
+    float fresnel = 0.5f * pow(1.0f + dotPr, 5);
+    if (fresnel > 0.5) {
+        return shade(reflectedRay, rayTrace(reflectedRay));
+    }
+    else {
+        return shade(refractedRay, rayTrace(refractedRay));
+    }
+    return fresnel * shade(reflectedRay, rayTrace(reflectedRay)) + (1 - fresnel) * shade(refractedRay, rayTrace(refractedRay));
```

Comment:
Hm, on a second glance do you even get to this combination, because you have an if/else before it?

Reply:
Ah, yes, this was commented out when making the actual scenes for the homework, I left it for demonstrative purposes, will remove it